### PR TITLE
8318150: StaticProxySelector.select should not throw NullPointerExceptions

### PR DESCRIPTION
**Pull request by djelinski**
`ProxySelector.select` and `connectFailed` methods should throw `IllegalArgumentException`s on null parameters.

Additionally, the `select` method doesn't need to be synchronized - the selector is immutable.

Tier2 green.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8318164](https://bugs.openjdk.org/browse/JDK-8318164) to be approved

### Issues
 * [JDK-8318150](https://bugs.openjdk.org/browse/JDK-8318150): StaticProxySelector.select should not throw NullPointerExceptions (**Bug** - P4)
 * [JDK-8318164](https://bugs.openjdk.org/browse/JDK-8318164): StaticProxySelector.select should not throw NullPointerExceptions (**CSR**)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [6ed6a983](https://git.openjdk.org/jdk/pull/16199/files/6ed6a983ac4c63a38bc87ac24b157996a1b146c6)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**) ⚠️ Review applies to [6ed6a983](https://git.openjdk.org/jdk/pull/16199/files/6ed6a983ac4c63a38bc87ac24b157996a1b146c6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16199/head:pull/16199` \
`$ git checkout pull/16199`

Update a local copy of the PR: \
`$ git checkout pull/16199` \
`$ git pull https://git.openjdk.org/jdk.git pull/16199/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16199`

View PR using the GUI difftool: \
`$ git pr show -t 16199`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16199.diff">https://git.openjdk.org/jdk/pull/16199.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16199#issuecomment-1764182987)

### CODE REVIEW

**Review by jaikiran — APPROVED** [review]
The change to the source and the migration of the test to junit looks fine to me.

**Review by dfuch — APPROVED** [review]
Looks reasonable to me. I believe CSR and release notes would be warranted. Also it would be good to double check whether other implementations of ProxySelector - e.g. when using -Djava.net.useSystemProxies=true also respect the contract of throwing IAE in those cases. We can't check that with a non-regression test unfortunately (unless it's a manual test) - but some code inspection might reveal whether the contract is respected or not.

**Comment by djelinski** [comment]
There are only 2 implementations of ProxySelector in the JDK, and they are both covered by this test. `-Djava.net.useSystemProxies=true` changes the behavior of DefaultProxySelector, but parameter validation code path is the same in both cases. I could add a separate `@run` line with `-Djava.net.useSystemProxies=true`, but in the current state it would just run the same code again.

Will post a CSR and a release note soon.

File: src/java.base/share/classes/java/net/ProxySelector.java [hunk 1]
```diff
@@ -207,12 +208,22 @@ static class StaticProxySelector extends ProxySelector {
...
+        public List<Proxy> select(URI uri) {
+            if (uri == null) {
+                throw new IllegalArgumentException("URI can't be null");
+            }
+            String scheme = uri.getScheme();
+            if (scheme == null) {
+                throw new IllegalArgumentException("protocol can't be null");
+            }
+            scheme = scheme.toLowerCase(Locale.ROOT);
             if (scheme.equals("http") || scheme.equals("https")) {
```
Comment (schlosna) [inline src/java.base/share/classes/java/net/ProxySelector.java:227]: maybe outside the scope of this PR, but could avoid the `toLowerCase`

```suggestion
            if (scheme.equalsIgnoreCase("http") || scheme.equalsIgnoreCase("https")) {
```

Comment (djelinski) [reply src/java.base/share/classes/java/net/ProxySelector.java]: Done.

**Pull request closed without merging**